doc: add `make html-live` and `html-live-fast`

JordanYates opened the pull request:
Add two new commands that automatically locally host the generated documentation upon completion and rebuild/rehost when the input files change without any user interaction.

For more info see: https://pypi.org/project/sphinx-autobuild/

JordanYates committed:
doc: add `make html-live` and `html-live-fast`

Add two new commands that automatically locally host the generated
documentation upon completion and rebuild/rehost when the input files
change without any user interaction.

For more info see: https://pypi.org/project/sphinx-autobuild/

Signed-off-by: Jordan Yates <jordan@embeint.com>

gmarull reviewed (commented):
does this work for rst files out of doc/ folder, eg board/samples docs?

JordanYates commented:
> does this work for rst files out of doc/ folder, eg board/samples docs?

It doesn't automatically pick them up, but they can be added with additional `--watch` parameters

aescolar reviewed (commented):
No complaints from me, but @gmarull and @kartben, and probably @carlescufi should  approve.
Thanks for remembering about the release notes :)

JordanYates commented:
> > > does this work for rst files out of doc/ folder, eg board/samples docs?
> > 
> > 
> > It doesn't automatically pick them up, but they can be added with additional `--watch` parameters
> 
> can you try?

`--watch` only takes folders, not patterns, so all I could really do is add `zephyr/boards` and `zephyr/samples` to the watch list. While that would work, it triggers a rebuild no matter which files are changes (`.dtsi`, `.yaml`, etc).

kartben commented:
> > > > does this work for rst files out of doc/ folder, eg board/samples docs?
> > > 
> > > 
> > > It doesn't automatically pick them up, but they can be added with additional `--watch` parameters
> > 
> > 
> > can you try?
> 
> `--watch` only takes folders, not patterns, so all I could really do is add `zephyr/boards` and `zephyr/samples` to the watch list. While that would work, it triggers a rebuild no matter which files are changes (`.dtsi`, `.yaml`, etc).

It might be possible to use `--ignore` or `re-ignore` to exclude some file patterns but it looks as if it would be better if there was a pattern-based way to clearly indicate the things to watch in the first place.

JordanYates committed:
doc: conf.py: build directory from environment

Get the output directory from environment variables instead of
attempting to reparse the CLI arguments, as these can actually refer
to a program other than `sphinx-build` (for example `sphinx-autobuild`).

Signed-off-by: Jordan Yates <jordan@embeint.com>

JordanYates commented:
Reparsing the argument list in `conf.py` no longer works properly when being run through `sphinx-autobuild`, updated to get the one variable of interest from the environment instead.

carlescufi commented:
> Reparsing the argument list in `conf.py` no longer works properly when being run through `sphinx-autobuild`, updated to get the one variable of interest from the environment instead.

Seems the doc build is failing @JordanYates

JordanYates commented:
> > Reparsing the argument list in `conf.py` no longer works properly when being run through `sphinx-autobuild`, updated to get the one variable of interest from the environment instead.
> 
> Seems the doc build is failing @JordanYates

Should be resolved, `ZEPHYR_BUILD` was the html dir, not the root build dir.

JordanYates commented:
Rebased on `v3.7.0`

kartben commented on doc/releases/release-notes-3.7.rst:
```diff
@@ -556,6 +556,13 @@ Build system and Infrastructure
+Documentation
+*************
+
+ * Added two new build commands, ``make html-live`` and ``make html-live-fast``, that automatically locally
+   host the generated documentation. They also automatically rebuild and rehost the documentation when changes
+   to the input ``.rst`` files are detected on the filesystem.
```
needs to move to new release notes :)
